packages/fxa-payments-server/server/lib/routes/legal-docs.js

```diff
@@ -0,0 +1,184 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+const config = require('../../config');
+const joi = require('joi');
+const fetch = require('node-fetch');
+const Sentry = require('@sentry/node');
+const { cdnFqdn, resultsCacheLimit } = config.get('legalDocLinks');
+const acceptLanuageParser = require('accept-language-parser');
+const supportedLanguages = require('fxa-shared/l10n/supportedLanguages.json');
+const DEFAULT_LOCALE = 'en';
+const CACHE_DURATION = 24 * 60 * 60; // a day in seconds
+const LANGUAGES_LIMIT = 3;
+
+/**
+ * A map for caching.  The key is a url and the value is an object with two properties:
+ *  - checkedOn: timestamp (in ms) when the last check against the url was performed
+ *  - result: boolean
+ *
+ * The size of this map will also be limited.  There are 71 supported languages
+ * and two docs per product (not price/plan).  With some generous rounding, we
+ * have 200 per product.  The current default is set to 65536, which should
+ * give us a decent runway.
+ */
+const cache = new Map();
+
+const getFullPath = (pathWithoutLocale, language) =>
+  // minimal handling of an errant '/'
+  pathWithoutLocale.endsWith('/')
+    ? `${pathWithoutLocale.substring(
+        0,
+        pathWithoutLocale.length - 1
+      )}.${language}.pdf`
+    : `${pathWithoutLocale}.${language}.pdf`;
+
+/**
+ * @param {obj} lang A language object from accept-language-parser
+ */
+const getLanguageTag = (lang) =>
+  lang.region ? `${lang.code}-${lang.region}` : lang.code;
+
+const isSupportedLanguage = (tag) => supportedLanguages.includes(tag);
+
+const isCacheExpired = (lastChecked) =>
+  Math.round((Date.now() - lastChecked) / 1000) > CACHE_DURATION;
+
+/**
+ * Keeping this simple: if the cache's full, delete half of the entries.
+ * It mutates the map from the arguments directly.
+ */
+const checkAndShrinkCache = (cacheMap, limit) => {
+  if (limit > 0 && cacheMap.size >= limit) {
+    const half = Math.ceil(limit / 2);
+    const keys = Array.from(cacheMap.keys()).slice(0, half);
+    for (const k of keys) {
+      cacheMap.delete(k);
+    }
+  }
+};
+
+const isValidDocUrl = async (url) => {
+  // Check cache first
+  const cachedVal = cache.get(url);
+  if (cachedVal !== undefined) {
+    if (isCacheExpired(cachedVal['checkedOn'])) {
+      cache.delete(url);
+    } else {
+      return cachedVal['result'];
+    }
+  }
+
+  checkAndShrinkCache(cache, resultsCacheLimit);
```

Comment:
hmm, I suppose so.  Since I stated above that the number of entries in the cache is not an exact science.  This prevents the number of entries ever going above the max.  Once per request would temporarily allow the number of entries to go above the max.

Reply:
Ah if the Map is at max when we get a new request, then we'd have max + 1 briefly, I guess? Yeah; I don't think it's a big deal either way. Makes sense not to exceed a max though if one is set.